### DISCUSSION TITLE
Q: is it possible to define dependencies generators in spec file?

### DISCUSSION THREAD

**Discussion by kloczek** [rpm-software-management/rpm#2485] [General]
According to https://rpm-software-management.github.io/rpm/manual/dependency_generators.html it is possible to define in .ttr files dependencies generators.
That works.

As long as .attr files defines macros I've been thinking that this will mean that such macros meybe is possible to define directly in spec file like
```spec
%define  __<package-name>_requires   %(<command>)
%define  __<package-name>_provides   %(<command>)
```
and looks like such macro is not used.

Typical use case is for example API version of just build foo-devel package as:
```spec
Provides: foo(<API.version>)
```
or for example package with X server provides ABI imterface
```spec
Provides: xserver-abi(<version>) = <ver.sion>
```
which is now defined manually for example in X server spec file by
```
%define         ansic_major     0
%define         ansic_minor     4
%define         videodrv_major  25
%define         videodrv_minor  2
%define         xinput_major    24
%define         xinput_minor    4
%define         extension_major 10
%define         extension_minor 0

Provides:       xserver-abi(ansic-%{ansic_major})               = %{ansic_minor}
Provides:       xserver-abi(extension-%{extension_major})       = %{extension_minor}
Provides:       xserver-abi(videodrv-%{videodrv_major})         = %{videodrv_minor}
Provides:       xserver-abi(xinput-%{xinput_major})             = %{xinput_minor}
```
And theoretically this could be generated by parse content of the files installed in buildroot 
```spec
%define __xorg-x11-server-Xorg_provides   %(<command>)
```
If such ABI version will change it will be necessary to update that manually but first someone will need to notie that all X server modules needs to be rebuilded because ABI has changed and now it fails on X Server start and rpm was not able to raise automatically broken dependencies.
Use cases for scenario simillar to above is much more.

Currently to have something like that it is neccessary to have inatall in `%_fileattrsdir` some `foo.attr` file and IMO it would be way better to have possiblity to define for exact in exact package spec file `%__<NAME>_provides` and other.

Issue is that as long as such macro is uaees and place where iit is defined are located in two places (.attr and spec file) it is nothing more than just asking for troubles. Other thinng is that as long as it is only dependecy henerator used only in one place (fro ecample provides generator) it would be good to have possibility to define that straight is spec file.

..or maybe it is possible to define something like that straight is spec? 🤔

**Comment by pmatilai** [comment]
It's not possible to define dependency generators without having an .attr file in the designated directory.
It would a handy feature, and the intention always was to add it at some point, just never gotten done 🫤

**Comment by kloczek** [comment]
That is OK (do I was not wrong) .. what do you think about create RFE? 🤔 

IMO especially such dependecy generator for `provides` would be usefull. IMO it would be godd to be able define different dependencies generators per sunbpackage so this is why I've been thinking about something like `%define  __<package-name>_{provides,requires}`.

**Comment by pmatilai** [comment]
The issue is discovery, ie which attributes to look for. That's what the .attr filename is used for, and there's no equivalent in the macro space. Otherwise, there's absolutely nothing special about the macros defined in the .attr files, and a regular old %define would do just fine for that.

**Comment by pmatilai** [comment]
Oh, and there already is an RFE for this: #2392
